### ISSUE TITLE
[Question] Parent catching forked tasks errors?

### ISSUE THREAD

**Issue by slorber** [yelouafi/redux-saga#543]
I'd like to do something like this:

``` javascript
function* saga() {
  const forkedTasks = [...];

  try {
    // Wait for all forked tasks (is there a more idiomatic way? maybe `join` ?)
    yield forkedTasks.map(task => task.done);
  }
  catch (e) {
    yield put(displayError());
    throw e;
  }
}
```

But it does not work. The parent seems to be cancelled and never get the opportunity to catch the error.

> Errors from child tasks automatically bubble up to their parents. If any forked task raises an uncaught error, then the parent task will abort with the child Error, and the whole Parent's execution tree (i.e. forked tasks + the main task represented by the parent's body if it's still running) will be cancelled.

So, when using forked tasks, is there a way for the parent to catch a child error? And eventually decide weither or not to kill the whole tasks tree?

The following seems to work, but it feels a bit unnatural and error prone to me. It's quite easy when using `spawn` to get tricked and let tasks alive while they should dye.

``` javascript
function* saga() {
  const spawnedTasks = [...];

  try {
    yield spawnedTasks.map(task => task.done);
  }
  catch (e) {
    yield put(displayError());
    throw e;
  }
  finally {
    yield spawnedTasks.map(task => cancel(task));
  }
}
```

Any idea on how this "error interception" pattern (I often want to do that) can be more idiomatically implemented?

**Comment by Andarist** [comment]
`yield forkedTasks.map(task => task.done)` - probably `array` handling should be implemented for `join` effect, gonna give it a try in a minute

As of this 'error interception' - I think this might be the difference between `fork` and `spawn` themselves. With `spawn` your task is detached and it is acting as top level task. 

I've debugged the issue a little bit.

1. Your both scenarios are reaching this point - https://github.com/yelouafi/redux-saga/blob/339ea9a8753d0a60dd006f2d8b2b26e88d75505e/src/internal/proc.js#L254 but with `spawn` those yielded promises are getting rejected and therefore error is thrown into the saga
2. Both your scenarios reache also this https://github.com/yelouafi/redux-saga/blob/339ea9a8753d0a60dd006f2d8b2b26e88d75505e/src/internal/proc.js#L84 but with `fork` `tasks.length === 1` and with `spawn` it is `0`. So with `fork` effect is geting settled during this cancelation process and when promises from 1. are geting rejected it is already settled so it wont run this https://github.com/yelouafi/redux-saga/blob/339ea9a8753d0a60dd006f2d8b2b26e88d75505e/src/internal/proc.js#L273 but `spawn` was not settled yet so it will run through the check and receive this error
3. `.return()` is used there so you could add `finally` block to your first snippet and perform some logic there

But anyway I dont know if what you are trying to accomplish is feasible, both scenarios will end up with `stdChannel` closed so you wont be able to `take` anything anymore etc.

I think this is the problem with your joining method based on quite internal promises, because such snippet seems to work (parent being able to catch child error)
```
function* saga() {
	try {
		const task2 = yield fork(function* () {
			yield call(delay, 1000)
			throw new Error('test')
		})

		yield take('TEST')
	} catch (e) {
		throw e;
	}
}
```

**Comment by slorber** [comment]
Hmmm,

I don't know, how comes your code is capable of catching the error? Maybe the task promise only resolves after the parent has been killed, but if a parent is blocking on an effect the exception is reinjected before the parent being killed?

**Comment by yelouafi** [comment]
@slorber I don't recommend using the `done` promise inside Sagas as it can lead to subtle race conditions. 

You may need to use the `join` effect which catch the task termination synchronously. 

```js
yield forkedTasks.map(join);
```

When I was implementing the new fork model, `join` was trivially implemented on top of `.done` but this caused me a lot of problems as the promise didn't notify the termination until the next microtask (I susped you were experiencing the same type of issue). That's why I implemented `join` on top of simple continuations (callbacks).

The main reason I left the `done` promise  in the lib is to be used by external functions (outside of Sagas).

>So, when using forked tasks, is there a way for the parent to catch a child error? And eventually decide weither or not to kill the whole tasks tree?

The short answer is no. But I think your case make perfect sense. The main reason for a parent being unable to catch errors from child tasks it that both evolve in parallel. But if at a moment a parent joined a child task, I think it makes sense at this point to let the parent catch the child error (since it's now blocked/synchronized with its child). 

so I think what we should do in this case is that when a parent joins a forked task, we should remove the child from the parent's [fork queue](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L46-L98). (But this leaves use with some questions especially regarding cancellation, it a `join` is cancelled shoudl we re-add the child to the fork queue).

>I don't know, how comes your last code snippet is capable of catching the error?

I dont think it'll work; the error form the forked task should abort the parent as well

**Comment by slorber** [comment]
Ok thanks :)

So in the end do you recommend any workaround in the meantime of a better solution, or should I continue temporarily to use `spawn`to achieve this result?

**Comment by yelouafi** [comment]
You can wrap your forks with the desired Error handler

```js
// Wrap forks with an Error handler
function* wrap(fn, args) {
  try {
    yield call(fn, ...args)
  }
  catch (e) {
    yield put(displayError());
    throw e;
  }
}

// a custom Effect creator for it
const forkWithErrHandler = (fn, ...args) => fork(wrap, fn, args)

function* saga() {
  //const forkedTasks = [...];
  
  // To fork a new task instead of `yield fork(fn, ...args)` use the wrapper
  yield forkWithErrHandler(fn, ...args)

  // yield forkedTasks.map(join);
  // You don't have to wait for forked tasks, saga will terminate when all forked tasks terminate

}
```

**Comment by slorber** [comment]
hmmm yes that looks like a good alternative! thanks

**Comment by granmoe** [comment]
Would something like this work for centralized error handling in all sagas in an app?

function* rootSaga () {
  yield sagas.map(saga =>
    spawn(function* () {
      while (true) {
        try {
          yield call(saga)
        } catch (e) {
          console.log(e)
        }
      }
    })
  )
}
Or would it possibly cause complex bugs due to scheduling issues internal to redux-saga? I'm wondering if something could get messed up going from the catch back to the try. Assuming this all happens synchronously, there should be no problem, right? Or wrong?

**Comment by yelouafi** [comment]
closing this. I think more finer-grained error control is needed. will use #570 for tracking this
